src/ints/mouse_vmware.cpp

```diff
@@ -0,0 +1,208 @@
+
+// ***************************************************************************
+// VMware interface implementation
+// ***************************************************************************
+
+static inline void CmdGetVersion() {
+    reg_eax = 0; // TODO: should we respond with something resembling VMware? For now 0 seems OK
+    reg_ebx = VMW_MAGIC;
+}
+
+static inline void CmdAbsPointerData() {
+    reg_eax = buttons_vmw;
+    reg_ebx = scaled_x;
+    reg_ecx = scaled_y;
+    reg_edx = static_cast<uint32_t>((wheel >= 0) ? wheel : 0x100 + wheel);
+
+    wheel = 0;
+}
+
+static inline void CmdAbsPointerStatus() {
+    reg_eax = updated ? 4 : 0;
+    updated = false;
+}
+
+static inline void CmdAbsPointerCommand() {
+    switch (reg_ebx) {
+    case VMW_ABSPNT::ENABLE:
+        break; // can be safely ignored
+    case VMW_ABSPNT::RELATIVE:
+        mouse_vmware = false;
+        LOG_MSG("MOUSE (PS/2): VMware protocol disabled");
+        GFX_UpdateMouseState();
+        break;
+    case VMW_ABSPNT::ABSOLUTE:
+        mouse_vmware = true;
+        wheel = 0;
+        LOG_MSG("MOUSE (PS/2): VMware protocol enabled");
+        GFX_UpdateMouseState();
+        break;
+    default:
+        LOG_WARNING("Mouse: unimplemented VMware subcommand 0x%08x", reg_ebx);
+        break;
+    }
+}
+
+static uint16_t PortRead_VMW(io_port_t, io_width_t) {
+    if (reg_eax != VMW_MAGIC)
+        return 0;
+
+    switch (reg_cx) {
+    case VMW_CMD::GETVERSION:         CmdGetVersion();        break;
+    case VMW_CMD::ABSPOINTER_DATA:    CmdAbsPointerData();    break;
+    case VMW_CMD::ABSPOINTER_STATUS:  CmdAbsPointerStatus();  break;
+    case VMW_CMD::ABSPOINTER_COMMAND: CmdAbsPointerCommand(); break;
+    default:
+        LOG_WARNING("Mouse: unimplemented VMware command 0x%08x", reg_ecx);
+        break;
+    }
+
+    return reg_ax;
+}
+
+bool MouseVMW_NotifyMoved(int32_t x_abs, int32_t y_abs) {
+    float vmw_x, vmw_y;
+    if (mouse_video.fullscreen) {
+        // We have to maintain the diffs (offsets) between host and guest
+        // mouse positions; otherwise in case of clipped picture (like
+        // 4:3 screen displayed on 16:9 fullscreen mode) we could have
+        // an effect of 'sticky' borders if the user moves mouse outside
+        // of the guest display area
+
+        if (x_abs + offset_x < mouse_video.clip_x)
+                offset_x = static_cast<int16_t>(mouse_video.clip_x - x_abs);
+        else if (x_abs + offset_x >= mouse_video.res_x + mouse_video.clip_x)
+                offset_x = static_cast<int16_t>(mouse_video.res_x + mouse_video.clip_x - x_abs - 1);
+
+        if (y_abs + offset_y < mouse_video.clip_y)
+                offset_y = static_cast<int16_t>(mouse_video.clip_y - y_abs);
+        else if (y_abs + offset_y >= mouse_video.res_y + mouse_video.clip_y)
+                offset_y = static_cast<int16_t>(mouse_video.res_y + mouse_video.clip_y - y_abs - 1);
```

Comment:
Use lambda to get rid of code duplication for x and y.

Reply:
Done.